packages/lib-webpack/src/webpack/DynamicRemotePlugin.ts

```diff
@@ -0,0 +1,167 @@
+import {
+  extensionArraySchema,
+  remoteEntryScript,
+  remoteEntryCallback,
+} from '@openshift/dynamic-plugin-sdk';
+import type { Extension } from '@openshift/dynamic-plugin-sdk';
+import * as _ from 'lodash-es';
+import * as fs from 'fs';
+import * as glob from 'glob';
+import * as path from 'path';
+import * as webpack from 'webpack';
+import type { PluginBuildMetadata } from '../types/plugin';
+import { pluginBuildMetadataSchema } from '../yup-schemas';
+import { GenerateManifestPlugin } from './GenerateManifestPlugin';
+import { PatchContainerEntryPlugin } from './PatchContainerEntryPlugin';
+
+const parseJSONFile = (filePath: string) => JSON.parse(fs.readFileSync(filePath, 'utf-8'));
+
+const loadPluginMetadata = (fileName: string, baseDir = process.cwd()) => {
+  const filePath = path.resolve(baseDir, fileName);
+  return validatePluginMetadata(parseJSONFile(filePath));
+};
+
+const loadExtensions = (globPattern: string, baseDir = process.cwd()) => {
+  const filePaths = glob.sync(globPattern, { cwd: baseDir, absolute: true, nodir: true });
+  return _.flatMap(filePaths.map((filePath) => validateExtensions(parseJSONFile(filePath))));
+};
+
+const validatePluginMetadata = (obj: unknown): PluginBuildMetadata =>
+  pluginBuildMetadataSchema.strict(true).validateSync(obj);
+
+const validateExtensions = (obj: unknown): Extension[] =>
+  extensionArraySchema.strict(true).validateSync(obj);
+
+const getSharedModulesWithStrictSingletonConfig = (modules: string[]) =>
+  modules.reduce((acc, moduleRequest) => {
+    acc[moduleRequest] = {
+      // Enforce a single version of the shared module to be used at runtime
+      singleton: true,
+      // Prevent plugins from using a fallback version of the shared module
+      import: false,
+    };
+    return acc;
+  }, {} as WebpackSharedModules);
```

Comment:
I do like this in some fashion -- it's a bit more knowledge needed to read this cleanly but I think this is a good functional way to do this. 👍

Reply:
> it's a bit more knowledge needed to read this cleanly

Code readability is very important.

`Object.fromEntries` takes an iterable and expects each element to be a tuple `[key, value]`, returning an object with properties constructed from these elements. It's basically a reverse operation to `Object.entries`.

Honestly, I don't think this is much better, compared to plain and simple `modules.reduce` array-to-object transformation.

Reply:
Pure functional programming has always taken a leap, mentally, for me. Reducer is a mixture of imperative and functional. Personally I think the functional programming way is more elegant and straight to the point. But I don't think this is needed for this PR / ever if you don't want. Preference.

I spent a couple years in `clojurescript` so pure functional things look cleaner to me than otherwise... but I'm definitely okay with the `reducer` way of things.

Reply:
@karelhala Well, there are two concerns here, logical and performance.

For each of these concerns, the question is, does it make sense to return a new accumulator object for each processed array element?

I agree that the object spread based form is shorter and nicer, but can be less performant and I don't see any real benefit from not reusing the same accumulator object from logical perspective :smiley: